core/feed_protocol/opds.py

```diff
@@ -11,11 +13,17 @@
 from core.util.opds_writer import OPDSMessage
 
 
-def get_serializer(content_type: Optional[str]) -> OPDS1Serializer | OPDS2Serializer:
-    if content_type and "json" in content_type:
-        return OPDS2Serializer()
-    else:
-        return OPDS1Serializer()
+def get_serializer(
```

Comment:
This is closer, but I think we need to write some tests around this function. 

I did a little bit of testing, and I don't think iterating over a `MIMEAccept` puts the results in priority order. It uses the order that the Accept header was sent in. It also returns a `Tuple[str, int]` mime / quality. 

I think we probably want to use the [best_match method](https://werkzeug.palletsprojects.com/en/2.0.x/datastructures/#werkzeug.datastructures.Accept.best_match) of `MIMEAccept`, which does the kind of sorting / matching we want here.

It would be nice to see some tests that test a few cases here: 
- No mime passed in
- A single simple mime type
- A couple more complicated Accept headers like: 
  - `Accept: application/atom+xml;profile=opds-catalog;kind=acquisition;q=0.08, application/opds+json;q=0.9`
  - `Accept: application/atom+xml;profile=opds-catalog;kind=acquisition, application/opds+json;q=0.9`

Reply:
Ok, I didn't quite grasp the q-value sets the priority and not the order of appearance. I'll fix this.

Reply:
Can you add a test case for these Accept header? 
- `Accept: application/atom+xml;profile=opds-catalog;kind=acquisition;q=0.08, application/opds+json;q=0.9`
  - I'd expect an `application/opds+json` response in that case, but I believe the current implementation will response with `application/atom+xml`. 
- `Accept: application/atom+xml;profile=opds-catalog;kind=acquisition`
  - I'd expect an `application/atom+xml` response in that case, but I believe we will serve `application/opds+json`

Reply:
Well... that was embarrassing. Seems like the MIMEAccept datatypes sorting is based FIRST on specificity THEN on q-value.
Changed the method to always sort on q-value.

Reply:
Is there a reason you aren't using the best_match method I mentioned above here? It would save a lot of sorting etc and lets you set a default.

Reply:
The datatype could be a List, this is only used in a test case for now, but was done to future proof the method in case we needed to do generate feeds offline.

Reply:
@RishiDiwanTT I believe that taking the specificity into account is part of the usual Accept header content negotiation dance. It would be nice to be able to lean on the complex matching provided by Flask instead of trying to reimplement it ourselves.

For tests, can we just directly pass in a `MimeAccept` class, or use the (admittedly annoying) flask machinery to fake a request for testing so we can directly consume `MimeAccept`.

Reply:
Updated. This caused one test case to change because specificity magic switched around the best match.